runtime-common/loader.ts

```diff
@@ -0,0 +1,212 @@
+        state: "fetching",
+        deferred: new Deferred<T>(),
+      };
+      this.modules.set(moduleIdentifier, module);
+      return await this.fetchModule(moduleIdentifier, module);
+    }
+
+    switch (module.state) {
+      case "fetching":
+        return await (module.deferred as Deferred<T>).promise;
+      case "preparing":
+        return (await module.moduleInstancePromise) as T;
+      case "evaluated":
+        return module.moduleInstance as T;
+      case "broken":
+        throw module.exception;
+      case "registered":
+        return await this.evaluateModule(moduleIdentifier, module);
+      default:
+        throw assertNever(module);
+    }
+  }
+
+  clearCache() {
+    this.modules = new Map();
+  }
+
+  private resolveModule(moduleIdentifier: string): string {
+    let moduleURL = new URL(moduleIdentifier);
+    if (baseRealm.inRealm(moduleURL)) {
+      return new URL(moduleURL.pathname, this.realm.baseRealmURL).href;
+    }
+    return moduleIdentifier;
+  }
+
+  private async fetchModule<T extends object>(
+    moduleIdentifier: string,
+    module: FetchingModule
+  ): Promise<T> {
+    let src: string;
+    try {
+      src = await this.fetch(new URL(moduleIdentifier));
+    } catch (exception) {
+      this.modules.set(moduleIdentifier, {
+        state: "broken",
+        exception,
+      });
+      throw exception;
+    }
+    src = transformSync(src, {
+      plugins: [
+        [TransformModulesAmd, { noInterop: true, moduleId: moduleIdentifier }],
+      ],
+    })?.code!;
+
+    try {
+      // this local is here for the evals to see
+      // @ts-ignore
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      let define = this.registerModule.bind(this);
+      eval(src);
+    } catch (exception) {
+      this.modules.set(moduleIdentifier, {
+        state: "broken",
+        exception,
+      });
+      throw exception;
+    }
+    let deferred = module.deferred as Deferred<T>;
+    deferred.fulfill(this.load(moduleIdentifier));
+    return deferred.promise;
+  }
+
+  private async evaluateModule<T extends object>(
+    moduleIdentifier: string,
+    module: RegisteredModule
+  ): Promise<T> {
+    let moduleInstance = Object.create(null);
+    let deferredModuleInstance = new Deferred<T>();
+    this.modules.set(moduleIdentifier, {
+      state: "preparing",
+      implementation: module.implementation,
+      moduleInstancePromise: deferredModuleInstance.promise,
+    });
+
+    try {
+      let dependencies = await Promise.all(
+        module.dependencyList.map((dependencyIdentifier) => {
+          if (dependencyIdentifier === "exports") {
+            return moduleInstance;
+          } else {
+            let absIdentifier = new URL(dependencyIdentifier, moduleIdentifier)
+              .href;
+            return this.load(absIdentifier);
+          }
+        })
+      );
+      module.implementation(...dependencies);
+      deferredModuleInstance.fulfill(moduleInstance);
```

Comment:
This was a key change that I had to make to prevent modules in the "preparing" state from returning `moduleInstance`s that were empty objects. Instead I set the state variable for the "preparing" state to a promise for a `moduleInstance`, which is fulfilled here. But i think this probably breaks cyclic module loading.